internal/internal_saga.go

```diff
@@ -0,0 +1,190 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+// All code in this file is private to the package.
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+const sagaOptionsContextKey = "sagaOptions"
+
+func getSagaOptions(ctx Context) *SagaOptions {
+	so := ctx.Value(sagaOptionsContextKey)
+	if so == nil {
+		return nil
+	}
+	return so.(*SagaOptions)
+}
+
+func setSagaParametersIfNotExist(ctx Context) Context {
+	params := getSagaOptions(ctx)
+	var newParams SagaOptions
+	if params != nil {
+		newParams = *params
+	}
+
+	return WithValue(ctx, sagaOptionsContextKey, &newParams)
+}
+
+type compensationOp struct {
+	activity interface{}
+	args     []interface{}
+}
+
+// Saga implements the logic to execute compensation operations
+// https://en.wikipedia.org/wiki/Compensating_transaction that is
+// often required in Saga applications.
+type Saga struct {
+	ctx               Context
+	options           *SagaOptions
+	compensationStack []*compensationOp
+	once              sync.Once
+	cancelFunc        CancelFunc
+
+	// since the cancelFunc is non-blocking, we use a canceled flag and lock
+	// to ensure compensation steps aren't run after the cancelFunc is called.
+	canceled bool
```

Comment:
I ran into a strange edge case in one of my tests where some compensation steps were executed even after the Saga context was canceled. The activity would return a `canceled` error but for some reason the activity was still executed. I'm not sure if that's expected or not. It was so not consistent either because sometimes the activity wouldn't execute but sometimes it did.

It didn't seem like a great user experience so I added this flag to prevent `Compensate` from running after `Cancel` is called.